### PR TITLE
Group events into Direct/Binning/Tile/Resolve passes

### PR DESCRIPTION
**Pull request by wangra-google**

### CODE REVIEW

File: dive_core/command_hierarchy.cpp [hunk 1]
```diff
@@ -980,6 +985,60 @@ bool CommandHierarchyCreator::OnPacket(const IMemoryManager &mem_manager,
         }
                 */
     }
+    else if (opcode == CP_SET_MARKER)
+    {
+        PM4_CP_SET_MARKER packet;
+        DIVE_VERIFY(mem_manager.CopyMemory(&packet, submit_index, va_addr, sizeof(packet)));
+        // as mentioned in adreno_pm4.xml, only b0-b3 are considered when b8 is not set
+        DIVE_ASSERT((packet.u32All0 & 0x100) == 0);
+        a6xx_marker    marker = static_cast<a6xx_marker>(packet.u32All0 & 0xf);
```
[the remark this answers is not in the document]
Comment (wangra-google) [reply dive_core/command_hierarchy.cpp:994]: A6xx and A7xx share the same. if you look at the domain `CP_SET_MARKER`:
```
<domain name="CP_SET_MARKER" width="32" varset="chip" prefix="chip" variants="A6XX-">
```
`variants="A6XX-"` means this applies to A6xx and above

[the remark this answers is not in the document]
Comment (wangra-google) [reply dive_core/command_hierarchy.cpp:994]: `a6xx_marker` is a bit misleading, but similar naming is used in other fields, e.g. `a3xx_regid`

**Comment by shanminchao** [comment]
I noticed that when you click the new nodes (e.g. Direct Rendering Pass), they don't show any packets in the Commands tab. This behaves differently from other nodes in the view. For example, if you click on the "Submit" node, it shows all the PM4 packets involved in that submit. Should we add this to keep it consistent? Could be useful too.

This does mean you'll have to mess with the m_shared_node_ib_parent_stack

**Comment by wangra-google** [comment]
> I noticed that when you click the new nodes (e.g. Direct Rendering Pass), they don't show any packets in the Commands tab. This behaves differently from other nodes in the view. For example, if you click on the "Submit" node, it shows all the PM4 packets involved in that submit. Should we add this to keep it consistent? Could be useful too.
> 
> This does mean you'll have to mess with the m_shared_node_ib_parent_stack

Thanks for pointing this out, new change should add the pm4s when the new node is selected.

**Comment by shanminchao** [comment]
> I noticed that when you click the new nodes (e.g. Direct Rendering Pass), they don't show any packets in the Commands tab. This behaves differently from other nodes in the view. For example, if you click on the "Submit" node, it shows all the PM4 packets involved in that submit. Should we add this to keep it consistent? Could be useful too.
> 
> This does mean you'll have to mess with the m_shared_node_ib_parent_stack

On second thought, I was wrong, and m_shared_node_ib_parent_stack is the wrong way to approach it. That stack is used to deal with INDIRECT_BUFFER packets and stuff like that, where a hierarchy is needed within the PM4 packets. 

Your recent change using m_shared_node_ib_parent_stack actually caused some unintended consequences because it was the wrong approach (sorry for leading you down the wrong path!). In particular, if you look at the Submit node now, it no longer contains all PM4 packets for the entire submit - the ones that are now part of the m_render_marker_index are no longer part of the submit, because of the way m_shared_node_ib_parent_stack works.

**Comment by wangra-google** [comment]
> > I noticed that when you click the new nodes (e.g. Direct Rendering Pass), they don't show any packets in the Commands tab. This behaves differently from other nodes in the view. For example, if you click on the "Submit" node, it shows all the PM4 packets involved in that submit. Should we add this to keep it consistent? Could be useful too.
> > This does mean you'll have to mess with the m_shared_node_ib_parent_stack
> 
> On second thought, I was wrong, and m_shared_node_ib_parent_stack is the wrong way to approach it. That stack is used to deal with INDIRECT_BUFFER packets and stuff like that, where a hierarchy is needed within the PM4 packets.
> 
> Your recent change using m_shared_node_ib_parent_stack actually caused some unintended consequences because it was the wrong approach (sorry for leading you down the wrong path!). In particular, if you look at the Submit node now, it no longer contains all PM4 packets for the entire submit - the ones that are now part of the m_render_marker_index are no longer part of the submit, because of the way m_shared_node_ib_parent_stack works.

make sense, updated the fix

File: dive_core/command_hierarchy.cpp [hunk 2]
```diff
@@ -984,6 +989,70 @@ bool CommandHierarchyCreator::OnPacket(const IMemoryManager &mem_manager,
...
+        if (add_child)
+        {
+            m_render_marker_index = AddNode(NodeType::kRenderMarkerNode, desc, 0);
+            AddChild(CommandHierarchy::kAllEventTopology, parent_node_index, m_render_marker_index);
+        }
+    }
+
+    if (m_render_marker_index != kInvalidRenderMarkerIndex)
+    {
+        AddSharedChild(CommandHierarchy::kEngineTopology, m_render_marker_index, packet_node_index);
```
Comment (shanminchao) [inline dive_core/command_hierarchy.cpp:1049]: I believe this new "render marker index" is only ever applicable to the kAllEventTopology, right? So in that case, is it even necessary to do the AddSharedChild for the other topologies?

Comment (wangra-google) [reply dive_core/command_hierarchy.cpp]: fixed

**Comment by shanminchao** [comment]
I noticed you just did a "Merge branch 'main' into group_drawcalls_tile", and now this PR consists of 9 commits. This is probably because you "merged" instead of "rebased". Can you rebase it instead?

**Comment by wangra-google** [comment]
> I noticed you just did a "Merge branch 'main' into group_drawcalls_tile", and now this PR consists of 9 commits. This is probably because you "merged" instead of "rebased". Can you rebase it instead?

(I was clicking the button in github web UI to update the branch to the latest. ) I guess it doesn't really matter with 9 commits since they will be squashed into 1 when merging the PR back to main?

**Comment by shanminchao** [comment]
> > I noticed you just did a "Merge branch 'main' into group_drawcalls_tile", and now this PR consists of 9 commits. This is probably because you "merged" instead of "rebased". Can you rebase it instead?
> 
> (I was clicking the button in github web UI to update the branch to the latest. ) I guess it doesn't really matter with 9 commits since they will be squashed into 1 when merging the PR back to main?

I'm not sure if that's true. There's a merge happening here, and you can't really squash a merge.
Looks like this in gitk right now:

image

**Review by shanminchao — APPROVED** [review]

**Pull request merged**